File: core/src/net/brenig/pixelescape/render/overlay/TutorialOverlay.java
package net.brenig.pixelescape.render.overlay;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.ui.Button;
import com.badlogic.gdx.scenes.scene2d.ui.Container;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.Align;

import net.brenig.pixelescape.game.data.GameMode;
import net.brenig.pixelescape.game.data.constants.StyleNames;
import net.brenig.pixelescape.game.data.constants.Textures;
import net.brenig.pixelescape.lib.Reference;
import net.brenig.pixelescape.render.ui.general.AnimationDrawable;
import net.brenig.pixelescape.render.ui.general.SwipeTabbedStack;
import net.brenig.pixelescape.render.ui.general.VerticalSpacer;
import net.brenig.pixelescape.screen.GameScreen;

/**
 * Overlay that is displayed at the first start of any gamemode explaining controls and features to the player
 */
public class TutorialOverlay extends OverlayWithUi implements SwipeTabbedStack.IOverSwipeListener, SwipeTabbedStack.IElementChangedListener {

	private final SwipeTabbedStack stack;
	private float timer = Float.MIN_VALUE;
	private static final float timer_amount = 0.2F;
	private final Button buttonLeft;
	private final Button buttonRight;
	private final Table controls;

	public TutorialOverlay(final GameScreen screen) {
		super(screen);

		final GameMode gameMode = screen.getGameMode();

		//Table headLayout = stage.createHeadUiLayoutTable();
		Table contentTable = stage.createContentUiLayoutTable();
		contentTable.defaults().fill();

		final int contentSizeX = screen.world.getWorldWidth() - 40;
		final int contentSizeY = screen.world.getWorldHeight() - 40;
		stack = new SwipeTabbedStack();
		stack.setOverSwipeListener(this);
		stack.setElementChangedListener(this);
		stack.setCycle(false);
		stack.setAllowLeftOverdraw(false);

		stack.add(screen.world.getPlayer().getMovementController().createTutorialWindow(getSkin(), screen, contentSizeX, contentSizeY));

		final int maxLabelWidth = contentSizeX - 60;
		if(gameMode.itemsEnabled()) {
			Table table = new Table(getSkin());
			table.setBackground(Textures.BUTTON_UP);
			table.defaults().padBottom(20);
			{
				Table row1 = new Table();
				{
					Image image = new Image(new AnimationDrawable(screen.game.getGameAssets().getItemAnimatedBackground()));
					row1.add(image).size(64);
				}
				{
					Label lbl = new Label("Collect Items", getSkin(), StyleNames.LABEL_WHITE);
					lbl.setAlignment(Align.left);
					lbl.setColor(Color.PURPLE);

					row1.add(lbl).left();
				}
				{
					Image image = new Image(new AnimationDrawable(screen.game.getGameAssets().getItemAnimatedBackground()));
					row1.add(image).size(64);
				}
				table.add(row1).width(maxLabelWidth).center().fillX();
			}
			table.row();
			{
				Label lbl = new Label("to gain abilities and effects!", getSkin(), StyleNames.LABEL_WHITE);
				lbl.setColor(Color.PURPLE);
				lbl.setAlignment(Align.center);
				lbl.setWrap(true);
				lbl.pack();

				table.add(lbl).width(maxLabelWidth).center();
			}

			stack.add(table);
		}
		if(gameMode.abilitiesEnabled()) {

			Table table = new Table(getSkin());
			table.setBackground(Textures.BUTTON_UP);
			table.defaults().padBottom(20);

			{
				Label lbl = new Label("Use your abilities wisely!", getSkin(), StyleNames.LABEL_WHITE);
				lbl.setColor(Color.NAVY);
				lbl.setWrap(true);
				lbl.setAlignment(Align.center);
				lbl.pack();

				table.add(lbl).width(maxLabelWidth).center();
			}

			stack.add(table);
		}

		gameMode.createCustomTutorial(getSkin(), stack, contentSizeX, contentSizeY);

		contentTable.add(stack).center().fill().pad(20).width(contentSizeX);

		controls = new Table();

		buttonLeft = new Button(getSkin(), StyleNames.BUTTON_ARROW_LEFT);
		buttonRight = new Button(getSkin(), StyleNames.BUTTON_ARROW_RIGHT);

		buttonRight.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				if(stack.hasNextElement()) {
					stack.next();
				} else {
					onAfterLast();
				}
			}
		});
		buttonRight.padLeft(40).padRight(40);

		buttonLeft.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				if(stack.hasLastElement()) {
					stack.last();
				}
			}
		});
		buttonLeft.padLeft(40).padRight(40);

		TextButton btnMainMenu = new TextButton("Main Menu", screen.game.getSkin());
		btnMainMenu.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				screen.showMainMenu();

			}
		});

		controls.add(buttonLeft).padLeft(40);

		controls.add(new Container<TextButton>(btnMainMenu)).fillX().expandX().center();
		controls.add(buttonRight).padRight(40);

		contentTable.add(new VerticalSpacer());
		contentTable.row();
		contentTable.add(controls);

		onElementChanged(0);
	}


	@Override
	public void renderFirst(float delta) {
		//noinspection PointlessBooleanExpression,ConstantConditions
		if(Reference.SCREEN_TINT_STRENGTH > 0) {
			if(timer != Float.MIN_VALUE) {
				renderScreenTint(Reference.SCREEN_TINT_STRENGTH * (timer / timer_amount));
			} else {
				renderScreenTint(Reference.SCREEN_TINT_STRENGTH);
			}
		}
	}

	@Override
	public void render(float delta) {
		super.render(delta);
		if(timer != Float.MIN_VALUE) {
			if(timer < 0) {
				screen.setOverlay(new CountDownOverlay(screen));
			} else {
				timer -= delta;
			}
		}
	}

	@Override
	public boolean doesPauseGame() {
		return true;
	}

	@Override
	public void onAfterLast() {
		stack.swipeOut(false);
		screen.setOverlayInputProcessor(null);
		controls.addAction(Actions.parallel(Actions.moveBy(0, -90, timer_amount), Actions.alpha(0.2F, timer_amount)));
		timer = timer_amount;
		screen.game.userData.setTutorialSeen(screen.getGameMode(), true);
	}

	@Override
	public void onBeforeFirst() {

	}

	@Override
	public void onElementChanged(int newElement) {
		if(!stack.hasNextElement()) {
			buttonRight.getColor().r = 0;
			buttonRight.getColor().g = 1;
			buttonRight.getColor().b = 0;
		} else {
			buttonRight.getColor().r = 1;
			buttonRight.getColor().g = 1;
			buttonRight.getColor().b = 1;
		}

		if(!stack.hasLastElement()) {
			buttonLeft.setDisabled(true);
		} else {
			buttonLeft.setDisabled(false);
		}
	}
}

File: core/src/net/brenig/pixelescape/render/overlay/TutorialOverlay.kt
package net.brenig.pixelescape.render.overlay

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.Align
import net.brenig.pixelescape.game.data.constants.StyleNames
import net.brenig.pixelescape.game.data.constants.Textures
import net.brenig.pixelescape.lib.Reference
import net.brenig.pixelescape.render.ui.general.AnimationDrawable
import net.brenig.pixelescape.render.ui.general.SwipeTabbedStack
import net.brenig.pixelescape.render.ui.general.VerticalSpacer
import net.brenig.pixelescape.screen.GameScreen

/**
 * Overlay that is displayed at the first start of any gamemode explaining controls and features to the player
 */
class TutorialOverlay(screen: GameScreen) : OverlayWithUi(screen), SwipeTabbedStack.IOverSwipeListener, SwipeTabbedStack.IElementChangedListener {

    private val stack: SwipeTabbedStack
    private var timer = java.lang.Float.MIN_VALUE
    private val buttonLeft: Button
    private val buttonRight: Button
    private val controls: Table

    init {

        val gameMode = screen.gameMode

        //Table headLayout = stage.createHeadUiLayoutTable();
        val contentTable = stage.createContentUiLayoutTable()
        contentTable.defaults().fill()

        val contentSizeX = screen.world.worldWidth - 40
        val contentSizeY = screen.world.worldHeight - 40
        stack = SwipeTabbedStack()
        stack.setOverSwipeListener(this)
        stack.setElementChangedListener(this)
        stack.setCycle(false)
        stack.setAllowLeftOverdraw(false)

        stack.add(screen.world.player.movementController.createTutorialWindow(skin, screen, contentSizeX, contentSizeY))

        val maxLabelWidth = contentSizeX - 60
        if (gameMode.itemsEnabled()) {
            val table = Table(skin)
            table.setBackground(Textures.BUTTON_UP)
            table.defaults().padBottom(20f)
            run {
                val row1 = Table()
                run {
                    val image = Image(AnimationDrawable(screen.game.gameAssets.itemAnimatedBackground))
                    row1.add(image).size(64f)
                }
                run {
                    val lbl = Label("Collect Items", skin, StyleNames.LABEL_WHITE)
                    lbl.setAlignment(Align.left)
                    lbl.color = Color.PURPLE

                    row1.add(lbl).left()
                }
                run {
                    val image = Image(AnimationDrawable(screen.game.gameAssets.itemAnimatedBackground))
                    row1.add(image).size(64f)
                }
                table.add(row1).width(maxLabelWidth.toFloat()).center().fillX()
            }
            table.row()
            run {
                val lbl = Label("to gain abilities and effects!", skin, StyleNames.LABEL_WHITE)
                lbl.color = Color.PURPLE
                lbl.setAlignment(Align.center)
                lbl.setWrap(true)
                lbl.pack()

                table.add(lbl).width(maxLabelWidth.toFloat()).center()
            }

            stack.add(table)
        }
        if (gameMode.abilitiesEnabled()) {

            val table = Table(skin)
            table.setBackground(Textures.BUTTON_UP)
            table.defaults().padBottom(20f)

            run {
                val lbl = Label("Use your abilities wisely!", skin, StyleNames.LABEL_WHITE)
                lbl.color = Color.NAVY
                lbl.setWrap(true)
                lbl.setAlignment(Align.center)
                lbl.pack()

                table.add(lbl).width(maxLabelWidth.toFloat()).center()
            }

            stack.add(table)
        }

        gameMode.createCustomTutorial(skin, stack, contentSizeX, contentSizeY)

        contentTable.add(stack).center().fill().pad(20f).width(contentSizeX.toFloat())

        controls = Table()

        buttonLeft = Button(skin, StyleNames.BUTTON_ARROW_LEFT)
        buttonRight = Button(skin, StyleNames.BUTTON_ARROW_RIGHT)

        buttonRight.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                if (stack.hasNextElement()) {
                    stack.next()
                } else {
                    onAfterLast()
                }
            }
        })
        buttonRight.padLeft(40f).padRight(40f)

        buttonLeft.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                if (stack.hasLastElement()) {
                    stack.last()
                }
            }
        })
        buttonLeft.padLeft(40f).padRight(40f)

        val btnMainMenu = TextButton("Main Menu", screen.game.skin)
        btnMainMenu.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                screen.showMainMenu()

            }
        })

        controls.add(buttonLeft).padLeft(40f)

        controls.add(Container(btnMainMenu)).fillX().expandX().center()
        controls.add(buttonRight).padRight(40f)

        contentTable.add(VerticalSpacer())
        contentTable.row()
        contentTable.add(controls)

        onElementChanged(0)
    }


    override fun renderFirst(delta: Float) {
        @Suppress("ConstantConditionIf")
        if (Reference.SCREEN_TINT_STRENGTH > 0) {
            if (timer != java.lang.Float.MIN_VALUE) {
                renderScreenTint(Reference.SCREEN_TINT_STRENGTH * (timer / timer_amount))
            } else {
                renderScreenTint(Reference.SCREEN_TINT_STRENGTH)
            }
        }
    }

    override fun render(delta: Float) {
        super.render(delta)
        if (timer != java.lang.Float.MIN_VALUE) {
            if (timer < 0) {
                screen.setOverlay(CountDownOverlay(screen))
            } else {
                timer -= delta
            }
        }
    }

    override fun doesPauseGame(): Boolean {
        return true
    }

    override fun onAfterLast() {
        stack.swipeOut(false)
        screen.setOverlayInputProcessor(null)
        controls.addAction(Actions.parallel(Actions.moveBy(0f, -90f, timer_amount), Actions.alpha(0.2f, timer_amount)))
        timer = timer_amount
        screen.game.userData.setTutorialSeen(screen.gameMode, true)
    }

    override fun onBeforeFirst() {

    }

    override fun onElementChanged(newElement: Int) {
        if (!stack.hasNextElement()) {
            buttonRight.color.r = 0f
            buttonRight.color.g = 1f
            buttonRight.color.b = 0f
        } else {
            buttonRight.color.r = 1f
            buttonRight.color.g = 1f
            buttonRight.color.b = 1f
        }

        if (!stack.hasLastElement()) {
            buttonLeft.isDisabled = true
        } else {
            buttonLeft.isDisabled = false
        }
    }

    override fun switchToPausedOverlayOnFocusChange(): Boolean {
        return false
    }

    companion object {
        private const val timer_amount = 0.2f
    }
}
